extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java

```diff
@@ -98,220 +97,207 @@ public String fromBytes(byte[] bytes)
     this.kafkaProperties = Preconditions.checkNotNull(kafkaProperties, "kafkaProperties required");
-    executorService = MoreExecutors.listeningDecorator(Execs.singleThreaded(
-        "kafka-factory-" + kafkaTopic + "-%s",
-        Thread.MIN_PRIORITY
-    ));
     this.cacheManager = cacheManager;
     this.connectTimeout = connectTimeout;
     this.injective = injective;
+    this.consumerConnectorFactory = consumerConnectorFactory;
+    this.executorService = MoreExecutors.listeningDecorator(Execs.singleThreaded(
+            "kafka-factory-" + kafkaTopic + "-%s",
+            Thread.MIN_PRIORITY
+    ));
   }
 
-  public KafkaLookupExtractorFactory(
-      NamespaceExtractionCacheManager cacheManager,
-      String kafkaTopic,
-      Map<String, String> kafkaProperties
-  )
-  {
-    this(cacheManager, kafkaTopic, kafkaProperties, 0, false);
+  void init() {
+    map = cacheManager.getCacheMap(factoryId);
   }
 
-  public String getKafkaTopic()
+  @Override
+  public boolean start()
   {
-    return kafkaTopic;
-  }
 
-  public Map<String, String> getKafkaProperties()
-  {
-    return kafkaProperties;
-  }
+    if(!advanceState(State.NEW, State.STARTING)) {
+      LOG.warn("Already started, not starting again. Current state: [%s]", currentState.get());
+      return currentState.get().equals(State.RUNNING);
+    }
 
-  public long getConnectTimeout()
-  {
-    return connectTimeout;
-  }
+    // TODO: if current state is STARTING, block until a terminal state is reached?
 
-  public boolean isInjective()
-  {
-    return injective;
+    final Properties kafkaProperties = buildProperties();
+    LOG.debug("About to listen to topic [%s] with group.id [%s]", kafkaTopic, factoryId);
+
+    init();
+
+    future = executorService.submit(makeRunnable(kafkaProperties));
+
+    if (!awaitStart()) {
+      currentState.set(State.FAILED);
+      return false;
+    }
+
+    advanceState(State.STARTING, State.RUNNING);
+
+    return true;
   }
 
-  @Override
-  public boolean start()
-  {
-    synchronized (started) {
-      if (started.get()) {
-        LOG.warn("Already started, not starting again");
-        return started.get();
-      }
-      if (executorService.isShutdown()) {
-        LOG.warn("Already shut down, not starting again");
-        return false;
-      }
-      final Properties kafkaProperties = new Properties();
-      kafkaProperties.putAll(getKafkaProperties());
-      if (kafkaProperties.containsKey("group.id")) {
-        throw new IAE(
-            "Cannot set kafka property [group.id]. Property is randomly generated for you. Found [%s]",
-            kafkaProperties.getProperty("group.id")
-        );
+  private boolean awaitStart() {
+    final Stopwatch stopwatch = Stopwatch.createStarted();
+    try {
+      while (!startingReads.await(100, TimeUnit.MILLISECONDS) && connectTimeout > 0L) {
+        // Don't return until we have actually connected
```

Comment:
What does this comment mean? Is it relating to the `await` being called at least once even if `connectTimeout` is 0?